src/simulator_context.rs

```diff
@@ -0,0 +1,508 @@
+    }
+    pub fn ctx_id() -> SimID {
+        TX_CONTEXT_ID.with(|f| f.borrow().clone())
+    }
+    pub fn clean() {
+        TX_CONTEXT_ID.with(|f| *f.borrow_mut() = 0.into());
+        PROC_CONTEXT_ID.with(|f| *f.borrow_mut() = 0.into());
+    }
+
+    pub fn start_process<F: Send + 'static + FnOnce(SimID, ProcID) -> i8>(
+        &mut self,
+        fds: &[Fd],
+        func: F,
+    ) -> ProcID {
+        let parent_id = ProcInfo::id();
+        let id = self.process_id_count.next();
+        let process = ProcInfo {
+            parent_id: parent_id.clone(),
+            inherited_fds: fds.to_vec(),
+            ..Default::default()
+        };
+
+        self.processes.insert(id.clone(), process);
+        let ctx_id = SimContext::ctx_id();
+
+        fds.iter().all(|fd| {
+            self.move_pipe(fd, id.clone());
+            true
+        });
+        self.process_status
+            .push(ProcStatus::WaitSpawn(id.clone(), false));
+
+        let id2 = id.clone();
+        let join_handle = std::thread::spawn(move || {
+            SimContext::update_ctx_id(ctx_id.clone(), Some(id.clone()));
+            let code = func(ctx_id.clone(), id.clone());
+
+            let mut gd = GlobalData::locked();
+            let cur_sim = gd.get_tx_mut(&SimContext::ctx_id());
+            cur_sim.close_all(&id);
+            cur_sim.process_io(None);
+
+            code
+        });
+
+        self.process_mut(&id2).join_handle = Some(join_handle);
+
+        id2
+    }
+    pub fn pid() -> ProcID {
+        ProcInfo::id()
+    }
+    pub fn inherited_fds(&self) -> Vec<Fd> {
+        let process = self.process(&ProcInfo::id());
+        process.inherited_fds.clone()
+    }
+    fn process(&self, id: &ProcID) -> &ProcInfo {
+        self.processes
+            .get(id)
+            .unwrap_or_else(|| panic!("unknow process id: {:?}", id))
+    }
+    fn process_mut(&mut self, id: &ProcID) -> &mut ProcInfo {
+        self.processes
+            .get_mut(id)
+            .unwrap_or_else(|| panic!("unknow process id: {:?}", id))
+    }
+    pub fn max_proc_spawned(&self) -> bool {
+        u64::from(self.process_id_count.clone()) > MAX_PROCESSES_COUNT
+    }
+    pub fn has_proc(&self, id: &ProcID) -> bool {
+        self.processes.contains_key(id)
+    }
+    pub fn get_event(&self) -> Event {
+        self.process(&ProcInfo::id()).scheduler_event.clone()
+    }
+    pub fn exit(&mut self, id: &ProcID) -> Option<JoinHandle<i8>> {
+        self.process_io(None);
+
+        let process = self.process_mut(id);
+
+        process.join_handle.take()
+    }
+
+    fn process_io(&mut self, fd: Option<&Fd>) {
+        // println!("==status 1: {:?}", self.process_status);
+
+        let mut update_rw: Vec<(usize, usize, bool)> = Vec::<(usize, usize, bool)>::new(); // Vec<(Read, Write)>
+        for i in 0..self.process_status.len() {
+            if let Some((_pid, rfd, rlen, _rbuf)) = self.process_status[i].read_wait() {
+                if rlen != &0 {
```

Comment:
rlen can be zero. Any reading or writing zero bytes operation can block.

Reply:
This is just to refresh the read and write cache. The notification is executed in `notify_status`, and each status will be judged again.